### PR TITLE
test: add vulnerability scan job creation integration tests

### PR DESCRIPTION
**Pull request by erikgb**
## Description

Introducing envtest cluster integration tests that can run without bootstrapping a full K8s cluster. This is the first step towards replacing the existing kind-based integration tests with envtest integration tests + end-to-end (e2e) tests using kind as proposed in #335.

This first PR adds integration tests for vulnerability scan job creation, and will be followed up by a new PR containing similar tests for finished scan job reconciling.

I will also create a PR for e2e-testing using [kuttl](https://kuttl.dev/), which together with the new integration tests should allow us to delete the existing integration tests. At least it seems so to me - since they appear quite "thin" and full of TODOs.

The contributor docs might appear a bit confusing in the transition period, since we now will have two types of "integration tests", but I hope that is acceptable for a short period.

## Related issues
- Relates to #335, but does not close it

## Checklist
- [x] I've read the [guidelines for contributing](https://github.com/aquasecurity/trivy-operator/blob/main/CONTRIBUTING.md) to this repository.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy-operator/tree/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by erikgb** [comment]
@chen-keinan Maybe you can scan through this draft and tell me what you think before I add tests for more workload types. 🙏

**Comment by chen-keinan** [comment]
> @chen-keinan Maybe you can scan through this draft and tell me what you think before I add tests for more workload types. 🙏

the test case look ok , isn't the envtest nfra is the same we have today : [envtest](https://github.com/aquasecurity/trivy-operator/blob/main/itest/trivy-operator)?

**Comment by erikgb** [comment]
> isn't the envtest nfra is the same we have today : [envtest](https://github.com/aquasecurity/trivy-operator/blob/main/itest/trivy-operator)?

WDYM? The itests run with kind. Envtest spins up a simple envtest cluster (from within the test suite), with just etcd and api-server - nothing more, nothing less. Did you try it? It works very well IMO.

**Comment by chen-keinan** [comment]
> > isn't the envtest nfra is the same we have today : [envtest](https://github.com/aquasecurity/trivy-operator/blob/main/itest/trivy-operator)?
> 
> WDYM? The itests run with kind. Envtest spins up a simple envtest cluster (from within the test suite), with just etcd and api-server - nothing more, nothing less. Did you try it? It works very well IMO.

does it require `Kubebuilder` bin to run locally ?

**Comment by erikgb** [comment]
> > > isn't the envtest nfra is the same we have today : [envtest](https://github.com/aquasecurity/trivy-operator/blob/main/itest/trivy-operator)?
> > 
> > 
> > WDYM? The itests run with kind. Envtest spins up a simple envtest cluster (from within the test suite), with just etcd and api-server - nothing more, nothing less. Did you try it? It works very well IMO.
> 
> does it require `Kubebuilder` bin to run locally ?
> 
> ```
> ERROR	controller-runtime.test-env	unable to start the controlplane	{"tries": 0, "error": "fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"}
> ```

No requirement for the kubebuilder CLI, but you need to install envtest for running tests from a IDE or standalone: https://book.kubebuilder.io/reference/envtest.html#installation. The Makefile takes care of this to run tests in CI (or locally). Does `make test` work for you?

**Comment by chen-keinan** [comment]
> Does `make test` work for you?

yes it works and I like it.

**Comment by erikgb** [comment]
> couple of comments , need to check if the current tests in `itests can be migrated` and I assume you plan to put it under folder on the main project

I am not sure if I understand what you re writing here. 🤔 My suggested overall plan is described in https://github.com/aquasecurity/trivy-operator/issues/335 (last part).

File: pkg/vulnerabilityreport/suite_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	err := testEnv.Stop()
+	Expect(err).NotTo(HaveOccurred())
+})
+
+func loadResource(obj runtime.Object, filename string) error {
+	yamlFile, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return err
+	}
+	err = yaml.UnmarshalStrict(yamlFile, obj)
```
Comment (josedonizetti) [inline pkg/vulnerabilityreport/suite_test.go:135]:
```suggestion
	return yaml.UnmarshalStrict(yamlFile, obj)
```

File: pkg/vulnerabilityreport/workload_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			createdJob := &batchv1.Job{}
+
+			// We'll need to retry getting this newly created Job, given that creation may not immediately happen.
+			Eventually(func() error {
+				return k8sClient.Get(ctx, jobLookupKey, createdJob)
+			}, timeout, interval).Should(Succeed())
+
+			Expect(createdJob).Should(WithTransform(NormalizeUntestableScanJobFields, Dyqual(expectedJob)))
+		},
+		Entry("Should create a scan Job for CronJob", &batchv1.CronJob{}, "cronjob.yaml", "cronjob-scan-job.yaml"),
```
Comment (josedonizetti) [inline pkg/vulnerabilityreport/workload_controller_test.go:64]: @erikgb Can we rename the expected files? Maybe `expected-cronjob-scan.yaml`, at least for me it was not very intuitive to look into the `testdata` folder and see difference.

**Review by josedonizetti — COMMENTED** [review]
@erikgb This looks great. Thank you! Two minor comments.

File: pkg/vulnerabilityreport/workload_controller_test.go (hunk 2 above)
Comment (erikgb) [reply pkg/vulnerabilityreport/workload_controller_test.go:64]: I agree, but can we rather name them `<resource>-expected-scan.yaml`? I like to keep them next to the resource-file.

File: pkg/vulnerabilityreport/suite_test.go (hunk 1 above)
Comment (erikgb) [reply pkg/vulnerabilityreport/suite_test.go]: Fixed, thanks

File: pkg/vulnerabilityreport/workload_controller_test.go (hunk 2 above)
Comment (josedonizetti) [reply pkg/vulnerabilityreport/workload_controller_test.go:64]: 👍

Comment (erikgb) [reply pkg/vulnerabilityreport/workload_controller_test.go]: Renamed

**Comment by chen-keinan** [comment]
LGTM , in term of folder structure I would suggest to move it one level up , to be in project level, so we could extent the itest for other scanners and reuse suite code

**Comment by erikgb** [comment]
> I see this almost as a controller unit tests with mocks, e2e would be more close to what we have currently on itests.

I do not agree with you here. 😉  Not using any mocks in these tests, so they are definitely integration tests (of the controller)! I'll probably mock out the logs reader in my next PR to add integration tests for the scan job controller. But that is because I have to, since api-server nor etcd can provide me with pod logs.

I would also like to fix https://github.com/aquasecurity/trivy-operator/issues/370 before adding integration tests for the scan job controller, but for now I'll just have to wait for your reviews and merge. What's holding this up? Just time to review/merge, or is there something else? Please tell me! 😺

**Comment by chen-keinan** [comment]
> > in term of folder structure I would suggest to move it one level up , to be in project level, so we could extent the itest for other scanners and reuse suite code
> 
> If I don't misunderstand you, I am against this. We should ideally have one integration test per controller, each of them getting a separate envtest cluster. Reusing code is of course something we should do! 

Not sure I follow , there is a lot of `testdata/fixture` and some of `suite_test` code that can be reused why duplicate ?

**Comment by erikgb** [comment]
@josedonizetti I am not sure if I am able to figure out what causes the race, but this PR is introducing ansync testing, which is new in trivy-operator. So I suspect it to already be an issue in the code on main, and that this PR only detects it. But I have never seen this race, nor locally on my workstation or in CI - and I have rebased numberous times triggering a new CI run.

But looking at the stack-traces, I think this piece of code looks suspicious:

https://github.com/aquasecurity/trivy-operator/blob/main/pkg/vulnerabilityreport/controller.go#L79-L80

Setting up two watches against the same GR seems like a bad idea. It should be either `v1.CronJob` or `v1beta1.CronJob` right?

**Comment by erikgb** [comment]
> 
> Not sure I follow , there is a lot of `testdata/fixture` and some of `suite_test` code that can be reused why duplicate ?

I suggest to reuse when we can reuse. The bootstrapping can be reused, but that should be part of the production code and not the test code. Not sure if we will/should reuse the resources.

**Comment by erikgb** [comment]
> > > Not sure I follow , there is a lot of `testdata/fixture` and some of `suite_test` code that can be reused why duplicate ?
> > 
> > 
> > I suggest to reuse when we can reuse. The bootstrapping can be reused, but that should be part of the production code and not the test code. Not sure if we will/should reuse the resources.
> 
> not sure I understand , most of code in `suite_test` can be reused , do you agree on it ?

No

**Comment by josedonizetti** [comment]
> Setting up two watches against the same GR seems like a bad idea. It should be either `v1.CronJob` or `v1beta1.CronJob` right?

Fixing this did the trick, no data races happening anymore. :D

**Pull request merged**